design/security-model.md

```diff
@@ -0,0 +1,360 @@
+  verbs: [ "get", "list", "watch", "patch", "create", "update, "delete" ]
+- apiGroups: ["rook.io"]
+  resources: ["*"]
+  verbs: [ "*" ]
+---
+kind: RoleBinding
+apiVersion: rbac.authorization.k8s.io/v1beta1
+metadata:
+  name: rook-system
+  namespace: rook-system
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: rook-system
+subjects:
+- kind: ServiceAccount
+  name: rook-system
+  namespace: rook-system
+```
+
+Notably absent here are privileges to set other RBAC rules read cluster-wide secrets, etc. Because the admin created the `rook-system` namespace and service account they are free to set policies on them using PSP or namespace quotas. Similar to the ones defined [here](https://github.com/rook/rook/blob/master/Documentation/kubernetes.md#rbac-for-podsecuritypolicies)).
+
+Also note that while we use a `ClusterRole` for rook-system we only use a `RoleBinding` to grant it access to the `rook-system` namespace. It does not have cluster-wide privileges.
+
+When creating a Rook cluster the cluster admin will continue to define the namespace and cluster CRD as follows:
+
+```
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: mycluster
+---
+apiVersion: rook.io/v1alpha1
+kind: Cluster
+metadata:
+  name: myrookcluster
+  namespace: mycluster
+	...
+```
+
+In addition we will require that the cluster-admin define a service account and role binding as follows:
+
+```
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: rook-cluster
+  namespace: mycluster
+---
+kind: Role
+apiVersion: rbac.authorization.k8s.io/v1beta1
+metadata:
+  name: rook-ceph-osd
+  namespace: mycluster
+rules:
+- apiGroups: [""]
+  resources: ["configmaps"]
+  verbs: [ "get", "list", "watch", "create", "update", "delete" ]
+---
+kind: RoleBinding
+apiVersion: rbac.authorization.k8s.io/v1beta1
+metadata:
+  name: rook-cluster
+  namespace: mycluster
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: rook-system
+subjects:
+- kind: ServiceAccount
+  name: rook-system
+  namespace: rook-system
+---
+kind: RoleBinding
+apiVersion: rbac.authorization.k8s.io/v1beta1
+metadata:
+  name: rook-system
+  namespace: mycluster
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: rook-cluster
+  namespace: rook-cluster
+subjects:
+- kind: ServiceAccount
+  name: rook-cluster
+  namespace: mycluster
+```
+
+This will grant the `rook-system` service account access to the new namespace and also setup a least privileged service account `rook-cluster` to be used for pods in this namespace that need K8S api access.
+
+With this approach `rook-system` will only have access to namespaces nominated by the cluster admin. Also we will no longer create any service accounts or namespaces enabling admins to set stable policies and quotas.
+
+Also all rook pods except the rook operator pod should run using `rook-cluster` service account in the namespace they're in.
+
+### Supporting common namespaces
```

Comment:
Has anyone asked for the ability to support multiple clusters in the same namespace? I don't actually remember hearing of anyone running multiple Rook clusters on the same K8s cluster. If someone does deploy multiple Rook clusters, they have to ensure that they don't try to use the same devices. It would also reduce the simplicity of the some properties in the CRDs and other resources. For example, the storage class would need to define both the cluster name and the cluster namespace. And the pool, file, and object CRDs would need a property that indicates the name of the cluster to join. Before we make this change I would want to make sure it is a useful scenario.

Reply:
Well, it could be useful for i.e. offer different classes of storage based on different devcies but certainly is an advanced topic due to the more advanced config required. 
To prevent user error rook could implement a series of checks (i.e. for duplicated usage of the same devices)

Reply:
There are two issues here -- 1) whether we support multiple rook clusters on the same K8S cluster (which we do today) and 2) wether they can share a namespace.

For (2) I'm concerned that we are forcing cluster-admins to create a namespace every time. According to https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/ namespaces are a unit of resource quotas, policy and security too. By forcing admins to multiple namespaces that means they need to duplicate and reconcile policies across namespaces.

I agree that (1) is infrequent, but for (2) we should let the cluster admin decide where rook pods run. This means the rook operator should NOT assume that all pods in a namespace belong to a single rook cluster.